### DISCUSSION TITLE
Blazor CSLA 7.0  SQL Transactions Does anyone have a samples of how to implement them

### DISCUSSION THREAD

**Discussion by jhuerta3** [MarimerLLC/csla#3717] [Questions] [answered]
Hello all,

I was wondering if anyone might have a sample of how to implement SQL transcations. I have been reading some post (https://github.com/MarimerLLC/csla/discussions/2979) but im not sure Im following. Maybe it cant even be done like i have listed below. I dont want to create a all new Dal object just to use a Scope transaction. Currently I am using Transient.  Im not sure if that is needed. 

The code below is how I think it should be done on my Blazor razor page. 

    private async Task Submit()
    {
        var builder = WebApplication.CreateBuilder();
        string conStr = builder.Configuration.GetConnectionString("DB_Connection");

        using DBContextScope _context = new DBContextScope(conStr);
        using var transaction = _context.Database.BeginTransaction();

        OrderEdit objNewOrder = OrderEditPortal.Create(_context);
        objNewOrder = await LoadOrder(objNewOrder);

        try
        {
            if (objNewOrder.IsValid)
            {
                var result = await objNewOrder.SaveAsync();
                if (result.BrokenRulesCollection.Count > 0)
                {
                    CheckBrokenRules(result.BrokenRulesCollection);
                    transaction.Rollback();
                }
                else
                {
                    LineItemEdit objLineItems = await LineItemEditPortal.CreateAsync(_context);
                    objLineItems = (LineItemEdit )await LoadLineItem(objLineItems);

                    if (objLineItems.IsValid)
                    {
                        var result1 = await objLineItems.SaveAsync();
                        transaction.Commit();
                    }
                    else
                    {
                        CheckBrokenRules(objLineItems.BrokenRulesCollection);
                        transaction.Rollback();
                    }
                    NavigationManager.NavigateTo("Error");
                }
            }
            else
            {
                CheckBrokenRules(objNewOrder.BrokenRulesCollection);
                transaction.Rollback();
            }
        }
        catch (Exception ex)
        {
            ErrorMessage = $"Error: {ex.ToString()}";
            ErrorMessageType = "danger";
        }
}

**Comment by rockfordlhotka** [comment]
Any transactions must be managed by the data portal or by your data portal operation method in the root object. There are _never_ transactional behaviors at the UI layers.

When using a remote data portal, all data access occurs on the app server, so the interface and interface control layers aren't even on the same physical machine.

When using a local data portal, the data portal creates a DI scope to isolate the logical "server-side" operations from the logical "client-side" operations, so it should be practically impossible for the interface and interface control layers to impact behaviors of the logical server (even though it is all running on the same physical machine).

You should be able to mark the data portal operation method of the root object with the CSLA transactional attribute and have ADO.NET automatically manage the transaction/rollback for you. That's the simple solution that usually works.

```c#
    [Update]
    [Transactional(TransactionalTypes.TransactionScope)]
    private void Update([Inject] IProjectDal dal)
```

This tells the data portal to start/commit/rollback a transaction for all ADO.NET operations within the object graph on the logical server. In this case, the DAL presumably injects a database connection - and that connection will automatically enlist in the transaction.

```c#
    private readonly PTrackerContext db;

    public ProjectDal(PTrackerContext context)
    {
      db = context;
    }
```

If you have data access that needs to be transactional, and _other_ data access that isn't transactional (like an event log), then you may need to take matters into your own hands. In that case you should use named services via DI to get access to the database connection for your business data, and a different database connection for your event log.

In this case, you'll probably use DI mechanisms to ensure that the business data connection has a transaction attached.

**Reply by Chicagoan2016** [reply]
@rockfordlhotka  In my Csla 4.xx applications, I log errors in the Data Access layers when I call stored procs (ADO.NET) or in 'UI' layer I check for DataPortalException.
Do you think 'DataPortal_OnDataPortalException'  would help, for logging events, when I use ' [Transactional(TransactionalTypes.TransactionScope)]' ? 
I currently don't need it but I find it very interesting.

Kind Regards

**Reply by rockfordlhotka** [reply]
In CSLA 7+ there is a concept called an _interceptor_ that you can use in the data portal pipeline. Here's an example:

https://github.com/MarimerLLC/csla/blob/main/Samples/CustomActivator/CustomActivator/CustomIntercepter.cs

The `Complete` method is invoked after the data portal operation is complete, and if there's an exception it is in the args parameter.

**Comment by jhuerta3** [comment]
Thank you both of you @Chicagoan2016  and @rockfordlhotka for responding.

I have another question: Is there a way to add two Child list for a business object? Or do I need to add a flag to child business object to differentiate between child_1 and Child_2.  Both Child_1 and Child_2 have different business propeties. Also the transaction will need be verified for (Parent, Child_1, and Child_2) to ensure that the transaction can be commited or rolledback.  

I was thinking of creating a BusinessListBase but that can only reference one child list. Any ideas?

**Reply by rockfordlhotka** [reply]
I think what you want is a BusinessBase root with a collection of child1 and a collection of child2.

**Reply by jhuerta3** [reply]
Yes but im not sure how I would verify the child1 and child2 busines rules. These two collections need to be businessbase objects for me to check the rules right? I was initially thinking that but I wasnt sure how I was going write it up. I tried to create two List properties but that didnot work also. I got an error.

**Reply by hurcane** [reply]
Child collections should consist of the appropriate CSLA list class (BusinessListBase or BusinessBindingListBase) and the contents of the list are BusinessBase classes.

With respect to rules, I find that they fall under three different categories:
- Independent
- Dependent on parent properties
- Dependent on the entire list

### Independent
Independent rules are rules that have no dependencies, or the dependencies are all within the class. The Required rule is an example of one that has no dependencies. If you have two properties that define a range, the rule class that checks that the range is valid will have a dependency, but all variables in the rule come from the same class.

### Dependent on parent
When rules are dependent on parent properties, a technique to manage that is to add a property on the child class that mirrors the property from the parent class. The rule that runs within the child class is then independent, since its properties all come from the child class. This requires a rule in the parent class to handle when the mirrored property changes. The rule loops the child class and sets the property on each one. This is such a common scenario that my company has created a standard rule that takes the parameters of the child collection and the name of the property. The Execute method uses reflection to assign the property.

### Dependent on the list
An example of a rule that is dependent on the list is a "no duplicates" requirement. To know whether the requirement is being met, the entire collection needs to be evaluated.

In my work, we deal with assemblies of hose (i.e. a bill of materials) that have a "components" collection. Some components have properties that affect the length of the hose. These properties have to be summed and then applied to the hose component.

For duplicates, you can put an IsDuplicate property on the child class. The child class would also have a simple rule to validate that the IsDuplicate property is false. In the parent class, an event handler for the ChildChanged event determines evaluates the list and sets the IsDuplicate property on every component. We usually do this with a a Linq query and the Group operator.

For the length rule, the ChildChanged event evaluates whether the property that changed affects the hose length. If so, it sums up the values, finds the hose component, and sets the appropriate property.

With complex business requirements, the ChildChanged event handler can be potentially dealing with many requirements.

**Reply by rockfordlhotka** [reply]
Nice writeup @hurcane.

I think @jhuerta3 probably needs to put the rule in the parent class that contains the two collections, and trigger the rule to be run using the `ChildChanged` event.

**Reply by jhuerta3** [reply]
Thank you @hurcane 
Hello @rockfordlhotka can you give me a quick sample of the two collections and calling the childChange.  Unless the ProjectTracker has a sample already. I need to reread all the books after realizing the Transactional attribute existed. By the way are there any new books? I have the older ones and also the CSLA4.0 sets.  
To be honest I think im over complicating this as I always do. Im working on a solution now but I would be interested on how to create the two collections sample. I will close this questions after. 

Thank you.

**Comment by Chicagoan2016** [comment, marked as answer]
@jhuerta3  I think what you need is very similar to ProjectEdit.cs (parent object) with child list similar to 'Resources' property which is of type ProjectResources.
The code is available in the ProjectTracker sample.
I would still recommend reading Using Csla 4.0 books, I have been working with Csla for some years and I keep those books on my machine.

**Reply by jhuerta3** [reply]
Thank you @Chicagoan2016
